trollmoves/s3downloader.py

```diff
@@ -0,0 +1,371 @@
+        self.loop = True
+        self.queue = queue
+        self.config = config
+        self.subscribe_nameserver = subscribe_nameserver
+
+    def stop(self):
+        """Stops the file listener."""
+        LOGGER.debug("Entering stop in FileListener ...")
+        self.loop = False
+        self.queue.put(None)
+
+    def _check_and_put_message_to_queue(self, msg):
+        if not self.loop:
+            LOGGER.warning("Self.loop false in FileListener %s", self.loop)
+            return False
+
+        # Check if it is a relevant message:
+        if self.check_message(msg):
+            LOGGER.info("Put the message on the queue...")
+            LOGGER.debug("Message = " + str(msg))
+            self.queue.put(msg)
+            LOGGER.debug("After queue put.")
+        return True
+
+    def run(self):
+        """"Run main loop subscribing to new messages adding these to a message queue."""
+        LOGGER.debug("Entering run in FileListener ...")
+        if type(self.config["subscribe-topic"]) not in (tuple, list, set):
+            self.config["subscribe-topic"] = [self.config["subscribe-topic"]]
+        try:
+            if 'services' not in self.config:
+                self.config['services'] = ''
+            subscriber_addresses = None
+            if 'subscriber_addresses' in self.config:
+                subscriber_addresses = self.config['subscriber_addresses'].split(',')
+
+            with Subscribe(self.config['services'], self.config['subscribe-topic'],
+                           True, addresses=subscriber_addresses,
+                           nameserver=self.subscribe_nameserver) as subscr:
+
+                LOGGER.debug("Entering for loop subscr.recv")
+                for msg in subscr.recv(timeout=1):
+                    if not self._check_and_put_message_to_queue(msg):
+                        break
+
+        except KeyError as ke:
+            LOGGER.info("Some key error. probably in config: %s", str(ke))
+            raise
+        except KeyboardInterrupt:
+            LOGGER.info("Received keyboard interrupt. Shutting down.")
+            raise
+
+    def check_message(self, msg):
+        """Check the message is valid type for this scope."""
+        if not msg:
+            # LOGGER.debug("message is None")
+            return False
+        if msg.type not in ('file', 'collection', 'dataset'):
+            LOGGER.debug("message type is not valid %s", str(msg.type))
+            return False
+        return True
+
+
+class FilePublisher(Thread):
+
+    """A publisher for result files listening to a publish queue.
+    Publishes the files via posttroll"""
+
+    def __init__(self, queue, nameservers):
+        Thread.__init__(self)
+        self.loop = True
+        self.queue = queue
+        self.service_name = 's3downloader'
+        self.nameservers = nameservers
+
+    def stop(self):
+        """Stops the file publisher."""
+        self.loop = False
+        self.queue.put("STOP")
```

Comment:
This is used as sentinel for tests, right?

Reply:
No, the queue loop `self.queue.get():` will hang until it gets a message.
It's a bit ugly, but it's the only way I have found to avoid the loop hanging when you want to stop.

Reply:
One option is to set `timeout=1` and handle the `Empty` exception it raises when there's nothing to get: https://docs.python.org/3/library/queue.html#queue.Queue.get

Reply:
OK, working on a rewrite/refactor.

Reply:
Changed a bit now, but the tests are failing somewhere else. Don't know how to fix this.